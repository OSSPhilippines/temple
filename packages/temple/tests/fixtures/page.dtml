<link rel="import" type="component" href="./dollar.tml" />
<style>
  body {
    font-family: Arial, Helvetica, sans-serif;
  }
  h1, h2, h3, h4, h5, h6, p {
    margin: 0;
    padding: 0;
  }
</style>
<script form observe="disabled,on">
  import snippets from '@/modules/snippets/home';
  const url = '/';
  const title = 'Temple - The reactive web component template engine.';
  const description = 'Temple is a template engine hat generates web components and support reactivity.';
  const snippet1 = `//on server:
    import temple from '@ossph/temple/compiler';
    //make a temple compiler
    const compiler = temple();
    //import a temple document file
    const { document } = await compiler.import('./page.dtml');
    //render final HTML
    const results = document.render();
    //show final HTML
    console.log(results);`;
  const snippet2 = `<style>
      body { 
        background-color: #DA532C; 
        color: #EFEFEF; 
      }
      .title { text-align: center; }
    </style>
    <script>
      const title = 'Hello World';
    </script>
    <html>
    <head>
      <title>{title}</title>
    </head>
    <body>
      <h1 class="title">{title}</h1>
    </body>
    </html>`;
  const APP_DATA = 'window.__APP_DATA__ = { title, description, url };';
  const BUILD_ID = 'a1b2c3d4';
  const MODE = 'development';
</script>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>{title}</title>
    <meta name="description" content={description} />
    <meta property="og:title" content={title} />
    <meta property="og:image" content="https://ossphilippines.github.io/temple/temple-logo.png" />
    <meta property="og:url" content={`https://ossphilippines.github.io/temple${url}`} />
    <meta property="og:type" content="website" />
    <link rel="stylesheet" type="text/css" href="/temple/styles/theme.css" />
    <link rel="stylesheet" type="text/css" href={`/temple/build/${BUILD_ID}.css`} />
    
    <script src="https://cdn.jsdelivr.net/npm/highlightjs-line-numbers.js/dist/highlightjs-line-numbers.min.js"></script>
    <script>{APP_DATA}</script>
    <script>console.log('hello')</script>
    <script src={`/temple/build/${BUILD_ID}.js`}></script>
    <if true={MODE !== 'production'}>
      <script src="/dev.js"></script>
    </if>
  </head>
  <body class="dark panel with-head">
    <dollar>
      <div title={title}><span>$</span> Ok</div>
    </dollar>
    <p>{snippet1}</p>
    <p>{snippet2}</p>
    <div>{`
<style>
  body { 
    background-color: #DA532C; 
    color: #EFEFEF; 
  }
  .title { text-align: center; }
</style>
<script>
  const title = 'Hello World';
</script>
<html>
<head>
  <title>{title}</title>
  <meta name="description" content={description} />
  <meta property="og:title" content={title} />
  <meta property="og:image" content="https://ossphilippines.github.io/temple/temple-logo.png" />
  <link rel="stylesheet" type="text/css" href="/temple/styles/theme.css" />
  <script src="https://cdn.jsdelivr.net/npm/highlightjs-line-numbers.js/dist/highlightjs-line-numbers.min.js"></script>
  <script>{APP_DATA}</script>
  <script>alert('hello')</script>
  <if true={MODE !== 'production'}>
    <script src="/dev.js"></script>
  </if>
</head>
<body>
  <h1 class="title">{title}</h1>
</body>
</html>
`}</div>
  </body>
</html>
