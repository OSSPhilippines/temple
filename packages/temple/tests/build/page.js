"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./components/Header1");
require("./components/Paragraph");
//# sourceMappingURL=page.js.map