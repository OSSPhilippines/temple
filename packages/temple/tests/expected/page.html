<!DOCTYPE html><html><head>
    <title>Test Title</title>
  
<style>
  .title { color: red; }
  .description { background-color: blue; }
</style>
<script>window.__SERVER_PROPS__ = {"title":"This is Header 1","description":"This is Paragraph","value":0}</script>
<script>(()=>{var t={336:function(t,e,r){"use strict";var s=this&&this.__awaiter||function(t,e,r,s){return new(r||(r=Promise))((function(n,i){function o(t){try{u(s.next(t))}catch(t){i(t)}}function a(t){try{u(s.throw(t))}catch(t){i(t)}}function u(t){var e;t.done?n(t.value):(e=t.value,e instanceof r?e:new r((function(t){t(e)}))).then(o,a)}u((s=s.apply(t,e||[])).next())}))},n=this&&this.__importDefault||function(t){return t&&t.__esModule?t:{default:t}};Object.defineProperty(e,"__esModule",{value:!0});const i=n(r(824)),o=n(r(516)),a=n(r(725));class u{constructor(){this.listeners={},this.event={event:"idle",pattern:"idle",parameters:[]},this.regexp=[]}static makeQueue(){return new o.default}emit(t,...e){return s(this,void 0,void 0,(function*(){a.default.require("string"==typeof t,"Argument 1 expected String");const r=this.match(t);if(!Object.keys(r).length)return i.default.NOT_FOUND;const n=u.makeQueue();return Object.keys(r).forEach((t=>{const i=r[t],o=i.pattern;void 0!==this.listeners[o]&&(i.args=e,this.listeners[o].forEach((t=>{n.add(((...e)=>s(this,void 0,void 0,(function*(){return this.event=Object.assign({},i,t),yield t.callback(...e)}))),t.priority)})))})),yield n.run(...e)}))}emitSync(t,...e){a.default.require("string"==typeof t,"Argument 1 expected String");const r=this.match(t);if(!Object.keys(r).length)return i.default.NOT_FOUND;const s=u.makeQueue();return Object.keys(r).forEach((t=>{const n=r[t],i=n.pattern;void 0!==this.listeners[i]&&(n.args=e,this.listeners[i].forEach((t=>{s.add(((...e)=>(this.event=Object.assign({},n,t),t.callback(...e))),t.priority)})))})),s.runSync(...e)}inspect(t){a.default.require("string"==typeof t,"Argument 1 expected String");const e=this.match(t);if(!Object.keys(e).length)return[];const r=u.makeQueue();return Object.keys(e).forEach((t=>{const s=e[t].pattern;void 0!==this.listeners[s]&&this.listeners[s].forEach((t=>{r.add(t.callback,t.priority)}))})),r.tasks}match(t){a.default.require("string"==typeof t,"Argument 1 expected String");const e={};return void 0!==this.listeners[t]&&(e[t]={event:t,pattern:t,parameters:[]}),this.regexp.forEach((r=>{const s=new RegExp(r.substr(r.indexOf("/")+1,r.lastIndexOf("/")-1),r.substr(r.lastIndexOf("/")+1));let n,i;if(-1===s.flags.indexOf("g")){if(n=t.match(s),!n||!n.length)return;i=[],Array.isArray(n)&&(i=n.slice(),i.shift())}else{if(n=Array.from(t.matchAll(s)),!Array.isArray(n[0])||!n[0].length)return;i=n[0].slice(),i.shift()}e[r]={event:t,pattern:r,parameters:i}})),e}on(t,e,r=0){return a.default.require("string"==typeof t||Array.isArray(t)||t instanceof RegExp,"Argument 1 expected String|RegExp (String|RegExp)[]"),a.default.require("function"==typeof e,"Argument 2 expected Function"),a.default.require("number"==typeof r,"Argument 3 expected Number"),Array.isArray(t)?(t.forEach((t=>{this.on(t,e,r)})),this):(t instanceof RegExp&&(t=t.toString(),-1===this.regexp.indexOf(t)&&this.regexp.push(t)),void 0===this.listeners[t]&&(this.listeners[t]=[]),this.listeners[t].push({priority:r,callback:e}),this)}unbind(t,e){if(!t&&!e){for(let t in this.listeners)delete this.listeners[t];return this}const r=this.listeners[t];return e||void 0===r?(void 0!==r&&"function"==typeof e&&r.forEach(((s,n)=>{e===s.callback&&(r.splice(n,1),r.length||delete this.listeners[t])})),t||"function"!=typeof e||Object.keys(this.listeners).forEach((t=>{this.listeners[t].forEach(((r,s)=>{e===r.callback&&(this.listeners[t].splice(s,1),this.listeners[t].length||delete this.listeners[t])}))})),this):(delete this.listeners[t],this)}use(...t){for(let e=0;e<t.length;e++){const r=t[e];this.regexp=this.regexp.concat(r.regexp),this.regexp=this.regexp.filter(((t,e,r)=>r.indexOf(t)===e));for(const t in r.listeners){const e=r.listeners[t];for(let r=0;r<e.length;r++)this.on(t,e[r].callback,e[r].priority)}}return this}}e.default=u},725:(t,e)=>{"use strict";Object.defineProperty(e,"__esModule",{value:!0});class r extends Error{static for(t,...e){return e.forEach((function(e){t=t.replace("%s",e)})),new this(t)}static forErrorsFound(t){const e=new this("Invalid Parameters");return e.errors=t,e}static require(t,e,...r){if(!t){for(const t of r)e=e.replace("%s",t);throw new this(e)}}constructor(t,e=500){super(),this.errors={},this.start=0,this.end=0,this.message=t,this.name=this.constructor.name,this.code=e}withCode(t){return this.code=t,this}withPosition(t,e){return this.start=t,this.end=e,this}toJSON(){return{error:!0,code:this.code,message:this.message}}}e.default=r},65:function(t,e,r){"use strict";var s=this&&this.__awaiter||function(t,e,r,s){return new(r||(r=Promise))((function(n,i){function o(t){try{u(s.next(t))}catch(t){i(t)}}function a(t){try{u(s.throw(t))}catch(t){i(t)}}function u(t){var e;t.done?n(t.value):(e=t.value,e instanceof r?e:new r((function(t){t(e)}))).then(o,a)}u((s=s.apply(t,e||[])).next())}))},n=this&&this.__importDefault||function(t){return t&&t.__esModule?t:{default:t}};Object.defineProperty(e,"__esModule",{value:!0}),e.File=void 0;const i=n(r(725));e.default=class{get data(){return this._data}get length(){return Object.keys(this._data).length}set data(t){i.default.require((null==t?void 0:t.constructor)===Object,"Argument 1 expected Object"),this._data=t}constructor(t={}){this._data=t,this.withArgs=new a(this),this.withFormData=new l(this),this.withPath=new u(this),this.withQuery=new c(this)}each(...t){return s(this,void 0,void 0,(function*(){const e=t.pop();let r=this.get(...t);if(!r||Array.isArray(r)&&!r.length||"string"==typeof r&&!r.length||"object"==typeof r&&!Object.keys(r).length)return!1;for(let t in r)if(!1===(yield e(r[t],t)))return!1;return!0}))}get(...t){if(!t.length)return this._data;if(!this.has(...t))return;const e=t.pop();let r=this._data;return t.forEach((t=>r=r[t])),r[e]}has(...t){if(!t.length)return!1;let e=!0;const r=t.pop();let s=this._data;return t.forEach((t=>{e&&("object"==typeof s[t]?s=s[t]:e=!1)})),!(!e||void 0===s[r])}remove(...t){if(!t.length)return this;if(!this.has(...t))return this;const e=t.pop();let r=this._data;return t.forEach((t=>{r=r[t]})),delete r[e],this}set(...t){if(t.length<1)return this;if("object"==typeof t[0])return Object.keys(t[0]).forEach((e=>{this.set(e,t[0][e])})),this;const e=t.pop();let r=t.pop(),s=this._data;return t.forEach(((e,r)=>{null!==e&&""!==e||(t[r]=e=Object.keys(s).length),"object"!=typeof s[e]&&(s[e]={}),s=s[e]})),null!==r&&""!==r||(r=Object.keys(s).length),s[r]=e,s=this._data,t.forEach((t=>{const e=s[t];var r;!Array.isArray(e)&&h(e)?s[t]=function(t){const e=[],r=Object.keys(t);return r.sort(),r.forEach((function(r){e.push(t[r])})),e}(e):Array.isArray(e)&&!h(e)&&(s[t]=(r=e,Object.assign({},r))),s=s[t]})),this}};class o{constructor(t){this.data=t.data,this.name=t.name,this.type=t.type}}e.File=o;class a{constructor(t){this.hash=t}set(...t){if(t.length<1)return this.hash;let e=t.pop();"number"!=typeof e&&(t.push(e),e=0);let r=t.pop();"string"==typeof r&&(r=r.split(" "));let s,n=0,i=e,o=r.length;for(;i<o;i++){const e=r[i],a=e.indexOf("=");if("--"!==e.substr(0,2))if("-"!==e.substr(0,1))-1===a?e.length&&this._format(t,n++,e):this._format(t,e.substr(0,a),e.substr(a+1));else{if("="===e.substr(2,1)){this._format(t,e.substr(1,1),e.substr(3));continue}const n=e.substr(1);for(let e=0;e<n.length;e++)s=n[e],this._format(t,s,!0);i+1<o&&"-"!==r[i+1][0]&&(this._format(t,s,r[i+1],!0),i++)}else{if(-1===a){if(s=e.substr(2),i+1<o&&"-"!==r[i+1][0]){this._format(t,s,r[i+1]),i++;continue}this._format(t,s,!0);continue}this._format(t,e.substr(2,a-2),e.substr(a+1))}}return this.hash}_format(t,e,r,s){switch(!0){case"string"!=typeof r:break;case"true"===r:r=!0;break;case"false"===r:r=!1;break;case!isNaN(r)&&!isNaN(parseFloat(r)):r=parseFloat(r);break;case!isNaN(r)&&!isNaN(parseInt(r)):r=parseInt(r)}t.length&&(e=t.join(".")+"."+e),e=String(e);const n=this.hash.withPath;if(!n.has(e)||s)return n.set(e,r),this.hash;const i=n.get(e);return Array.isArray(i)?(i.push(r),n.set(e,i),this.hash):(n.set(e,[i,r]),this.hash)}}class u{constructor(t){this.hash=t}each(t,e,r="."){return s(this,void 0,void 0,(function*(){const s=t.split(r);return yield this.hash.each(...s,e)}))}get(t,e="."){const r=t.split(e);return this.hash.get(...r)}has(t,e="."){const r=t.split(e);return this.hash.has(...r)}remove(t,e="."){const r=t.split(e);return this.hash.remove(...r)}set(t,e,r="."){const s=t.split(r);return this.hash.set(...s,e)}}class c{constructor(t){this.hash=t}set(...t){if(t.length<1)return this.hash;const e=t.pop(),r="~~"+Math.floor(1e4*Math.random())+"~~";return e.split(/\&/gi).forEach((s=>{const[n,i]=s.split("=",2),o=n.replace(/\]\[/g,r).replace("[",r).replace(/\[/g,"").replace(/\]/g,"").split(r);o.map((t=>{const e=parseInt(t);return isNaN(e)||-1!==t.indexOf(".")?t:e}));const a=t.concat(o);if(/(^\{.*\}$)|(^\[.*\]$)/.test(i))try{return e.set(...a,JSON.parse(i))}catch(t){}isNaN(parseFloat(i))?"true"===i?this.hash.set(...a,!0):"false"===i?this.hash.set(...a,!1):"null"===i?this.hash.set(...a,null):this.hash.set(...a,i):this.hash.set(...a,parseFloat(i))})),this.hash}}class l{constructor(t){this.hash=t}set(...t){if(t.length<1)return this.hash;const e=t.pop(),r="string"==typeof e?Buffer.from(e):e,s=this._getBoundary(r);let n=[];for(let e=0;e<r.length;e++){const i=this._getLine(r,e);if(null===i)break;const o=i.buffer;if(0===o.toString().indexOf(s)){if(n.length&&this._setPart(t,this._getPart(n)),o.toString()===`${s}--`)break;n=[]}else n.push(o);e=i.i}return this.hash}_getBoundary(t){var e;const r=null===(e=this._getLine(t,0))||void 0===e?void 0:e.buffer;return r?r.slice(0,r.length-1).toString():null}_getLine(t,e){const r=[];for(;e<t.length;e++){const s=t[e];if(r.push(s),10===s||13===s)return{i:e,buffer:Buffer.from(r)}}return r.length?{i:e,buffer:Buffer.from(r)}:null}_getPart(t){var e;const r=[];do{r.push(null===(e=t.shift())||void 0===e?void 0:e.toString())}while(t.length&&(1!==t[0].length||10!==t[0][0]&&13!==t[0][0]));const s=t[t.length-1];t[t.length-1]=s.slice(0,s.length-1);const n=Buffer.concat(t.slice(1)),i={};for(const t of r)if(t&&-1!==t.indexOf(":")){const[e,r]=t.toString().split(":",2);i[e.trim().toLowerCase()]=r.trim()}const o={};return"string"==typeof i["content-disposition"]&&i["content-disposition"].split(";").forEach((t=>{const e=t.trim().match(/^([a-zA-Z0-9_\-]+)=["']([^"']+)["']$/);e&&e.length>2&&(o[e[1]]=e[2])})),{headers:i,body:n,form:o}}_setPart(t,e){if(!e.form.name)return this;const r="~~"+Math.floor(1e4*Math.random())+"~~",s=e.form.name.replace(/\]\[/g,r).replace("[",r).replace(/\[/g,"").replace(/\]/g,"").split(r);s.map((t=>{const e=parseInt(t);return isNaN(e)||-1!==t.indexOf(".")?t:e}));const n=t.concat(s);if(!e.form.filename){const t=e.body.toString();if(/(^\{.*\}$)|(^\[.*\]$)/.test(t))try{return this.hash.set(...n,JSON.parse(t))}catch(t){}return isNaN(parseFloat(t))?"true"===t?this.hash.set(...n,!0):"false"===t?this.hash.set(...n,!1):"null"===t?this.hash.set(...n,null):this.hash.set(...n,t):this.hash.set(...n,parseFloat(t)),this}this.hash.set(...n,new o({data:e.body,name:e.form.filename,type:e.headers["content-type"]}))}}function h(t){if("object"!=typeof t)return!1;const e=Object.keys(t).length;if(!e)return!1;for(let r=0;r<e;r++)if(void 0===t[r])return!1;return!0}},824:(t,e)=>{"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.default={get ABORT(){return{code:308,message:"Aborted"}},get ERROR(){return{code:500,message:"Internal Error"}},get NOT_FOUND(){return{code:404,message:"Not Found"}},get OK(){return{code:200,message:"OK"}}}},516:function(t,e,r){"use strict";var s=this&&this.__awaiter||function(t,e,r,s){return new(r||(r=Promise))((function(n,i){function o(t){try{u(s.next(t))}catch(t){i(t)}}function a(t){try{u(s.throw(t))}catch(t){i(t)}}function u(t){var e;t.done?n(t.value):(e=t.value,e instanceof r?e:new r((function(t){t(e)}))).then(o,a)}u((s=s.apply(t,e||[])).next())}))},n=this&&this.__importDefault||function(t){return t&&t.__esModule?t:{default:t}};Object.defineProperty(e,"__esModule",{value:!0});const i=n(r(824)),o=n(r(725));e.default=class{constructor(){this.tasks=[],this.lower=0,this.upper=0}get length(){return this.tasks.length}add(t,e=0){return o.default.require("function"==typeof t,"Argument 1 expected Function"),o.default.require("number"==typeof e,"Argument 2 expected Number"),e>this.upper?this.upper=e:e<this.lower&&(this.lower=e),this.tasks.push({callback:t,priority:e}),this.tasks.sort(((t,e)=>t.priority<=e.priority?1:-1)),this}push(t){return o.default.require("function"==typeof t,"Argument 1 expected Function"),this.add(t,this.lower-1)}shift(t){return o.default.require("function"==typeof t,"Argument 1 expected Function"),this.add(t,this.upper+1)}run(...t){return s(this,void 0,void 0,(function*(){if(!this.tasks.length)return i.default.NOT_FOUND;for(;this.tasks.length;){const e=this.tasks.shift();if(e&&!1===(yield e.callback(...t)))return i.default.ABORT}return i.default.OK}))}runSync(...t){if(!this.tasks.length)return i.default.NOT_FOUND;for(;this.tasks.length;)if(!1===this.tasks.shift().callback(...t))return i.default.ABORT;return i.default.OK}}},500:function(t,e,r){"use strict";var s=this&&this.__createBinding||(Object.create?function(t,e,r,s){void 0===s&&(s=r);var n=Object.getOwnPropertyDescriptor(e,r);n&&!("get"in n?!e.__esModule:n.writable||n.configurable)||(n={enumerable:!0,get:function(){return e[r]}}),Object.defineProperty(t,s,n)}:function(t,e,r,s){void 0===s&&(s=r),t[s]=e[r]}),n=this&&this.__setModuleDefault||(Object.create?function(t,e){Object.defineProperty(t,"default",{enumerable:!0,value:e})}:function(t,e){t.default=e}),i=this&&this.__importStar||function(t){if(t&&t.__esModule)return t;var e={};if(null!=t)for(var r in t)"default"!==r&&Object.prototype.hasOwnProperty.call(t,r)&&s(e,t,r);return n(e,t),e},o=this&&this.__importDefault||function(t){return t&&t.__esModule?t:{default:t}};Object.defineProperty(e,"__esModule",{value:!0}),e.TaskQueue=e.File=e.HashStore=e.Statuses=e.Exception=e.EventEmitter=void 0;const a=o(r(336));e.EventEmitter=a.default;const u=o(r(725));e.Exception=u.default;const c=o(r(824));e.Statuses=c.default;const l=i(r(65));e.HashStore=l.default,Object.defineProperty(e,"File",{enumerable:!0,get:function(){return l.File}});const h=o(r(516));e.TaskQueue=h.default},383:(t,e,r)=>{t.exports={...r(500)}},92:(t,e,r)=>{"use strict";Object.defineProperty(e,"__esModule",{value:!0});const s=r(383);class n extends s.Exception{static for(t,...e){return e.forEach((function(e){t=t.replace("%s",e)})),new this(t)}static forErrorsFound(t){const e=new this("Invalid Parameters");return e.errors=t,e}static require(t,e,...r){if(!t){for(const t of r)e=e.replace("%s",t);throw new this(e)}}}e.default=n},536:function(t,e,r){"use strict";var s=this&&this.__importDefault||function(t){return t&&t.__esModule?t:{default:t}};Object.defineProperty(e,"__esModule",{value:!0});const n=s(r(339)),i=s(r(565)),o=window.__SERVER_PROPS__;class a extends HTMLElement{static get current(){return a._current}get props(){return Object.assign({},this._element.attributes)}get initiated(){return this._initiated}set props(t){if(this._element.serialize(t)!==this._element.serialize()){const e=Object.assign({},t);for(const[r,s]of Object.entries(t))if(s.startsWith("data:")){let t=s.substring(5);"true"===s?t=!0:"false"===s?t=!1:"null"===s&&(t=null),e[r]=t}else if(s.startsWith("prop:")){const t=s.substring(5);void 0!==o[t]&&(e[r]=o[t])}this._element.setAttributes(e),this.render()}}constructor(){super(),this._shadowRoot=null,this._initiated=!1,this._template=null,this._element=n.default.register(this)}adoptedCallback(){this.render()}attributeChangedCallback(t,e,r){this.props=Object.assign(Object.assign({},this.props),{[t]:r})}connectedCallback(){this.wait()}disconnectedCallback(){}render(){a._current=this;const t=this.styles();this._template||(this._template=this.template());const e=this._template().filter(Boolean);if(0===t.length)this.textContent="",e.forEach((t=>this.appendChild(t)));else{this._shadowRoot||(this._shadowRoot=this.attachShadow({mode:"open"})),this.textContent="";const r=document.createElement("style");r.innerText=t,this._shadowRoot.appendChild(r),e.forEach((t=>{var e;return null===(e=this._shadowRoot)||void 0===e?void 0:e.appendChild(t)}))}a._current=null,this._initiated=!0,i.default.emit("render",this)}ready(){const t=[];let e=this;for(;e.parentNode;)e=e.parentNode,t.push(e);return[this,...t].some((t=>t.nextSibling))||"loading"!==document.readyState}update(t){const e=Object.fromEntries(Array.from(this.attributes).map((t=>[t.nodeName,t.nodeValue])));this.props=Object.assign(Object.assign(Object.assign({},this.props),e),{children:t})}wait(){if(this.ready())this.update(this.innerHTML);else{const t=new MutationObserver((()=>{this.ready()&&(this.update(this.innerHTML),t.disconnect())}));t.observe(this,{childList:!0})}}}a._current=null,e.default=a},339:function(t,e,r){"use strict";var s=this&&this.__importDefault||function(t){return t&&t.__esModule?t:{default:t}};Object.defineProperty(e,"__esModule",{value:!0});const n=s(r(565));class i{static create(t,e,r=[]){const s=document.createElement(t);for(const[t,r]of Object.entries(e))"string"==typeof r&&"children"!==t&&s.setAttribute(t,r);return r.forEach((t=>s.appendChild(t))),this.register(s,e)}static filter(t){const e=[];return this.registry.forEach(((r,s)=>{t(r,s)&&e.push(r)})),e}static get(t){return this.registry.get(t)||null}static map(t){const e=[];return this.registry.forEach(((r,s)=>{e.push(t(r,s))})),e}static register(t,e){return this.registry.has(t)?this.get(t):new i(t,e||{})}get attributes(){return Object.assign({},this._attributes)}get element(){return this._element}constructor(t,e){this._element=t,this._attributes=e,i.registry.set(this._element,this)}hasAttribute(t){return t in this._attributes}getAttribute(t){return this._attributes[t]}removeAttribute(t){const e=this.getAttribute(t);return void 0===e||(delete this._attributes[t],this._element.removeAttribute(t),n.default.emit("attribute-remove",this,t,e)),this}serialize(t){return t||(t=this._attributes),JSON.stringify(t,((t,e)=>"function"==typeof e?e.toString():e))}setAttribute(t,e){const r=this.getAttribute(t);return r===e||(this._attributes[t]=e,"string"==typeof e&&"children"!==t&&this._element.setAttribute(t,e),void 0===r?n.default.emit("attribute-create",this,t,e):n.default.emit("attribute-update",this,t,e,r)),this}setAttributes(t){for(const[e,r]of Object.entries(t))this.setAttribute(e,r);const e=Object.keys(t);for(const t of Object.keys(this._attributes))e.includes(t)||this.removeAttribute(t);return this}}i.registry=new Map,e.default=i},565:(t,e,r)=>{"use strict";Object.defineProperty(e,"__esModule",{value:!0}),e.TempleEmitter=void 0;const s=r(383);class n extends s.EventEmitter{on(t,e,r=0){if("ready"===t){const t=e;return"complete"===document.readyState||"interactive"===document.readyState?setTimeout(t,1):document.addEventListener("DOMContentLoaded",(()=>t())),this}return super.on(t,e,r)}}e.TempleEmitter=n;const i=new n;e.default=i},214:function(t,e,r){"use strict";var s=this&&this.__importDefault||function(t){return t&&t.__esModule?t:{default:t}};Object.defineProperty(e,"__esModule",{value:!0}),e.bindAttribute=void 0;const n=s(r(339)),i=s(r(565));function o(t,e){i.default.on("ready",(()=>{n.default.filter((e=>e.hasAttribute(t))).forEach(e)})),i.default.on("render",(r=>{Array.from(r.querySelectorAll("*")).forEach((r=>{const s=n.default.get(r);s&&s.hasAttribute(t)&&e(s)}))}))}e.bindAttribute=o,o("if",(t=>{const e=t.getAttribute("if");!1===e||"false"===e?t.element.remove():"function"!=typeof e||e()||t.element.remove()})),["click","dblclick","mousedown","mouseup","mousemove","mouseover","mouseout","wheel","keydown","keypress","keyup","blur","change","contextmenu","focus","input","submit","invalid","reset","search","select","copy","cut","paste","drag","dragstart","dragend","dragover","dragenter","dragleave","drop","scroll","durationchange","ended","error","loadeddata","loadedmetadata","loadstart","pause","play","playing","progress","ratechange","seeked","seeking","stalled","suspend","timeupdate","volumechange","waiting","animationstart","animationend","animationiteration","transitionend","toggle"].forEach((t=>o(t,(e=>{const r=e.getAttribute(t);"function"==typeof r&&e.element.addEventListener(t,r)}))))},357:function(t,e,r){"use strict";var s=this&&this.__importDefault||function(t){return t&&t.__esModule?t:{default:t}};Object.defineProperty(e,"__esModule",{value:!0});const n=s(r(536));e.default=function(){return n.default.current?n.default.current.props:{}}},975:function(t,e,r){"use strict";var s=this&&this.__importDefault||function(t){return t&&t.__esModule?t:{default:t}};Object.defineProperty(e,"__esModule",{value:!0}),e.SignalRegistry=void 0;const n=s(r(536)),i=s(r(92));class o{static observe(t,e){const r={raw:e};Object.defineProperty(r,"value",{get:()=>r.raw,set(e){const s=o.serialize(e)!==o.serialize(r.raw);r.raw=e,s&&t.render()}});const s=this.observers.get(t);return s?(s.observed++,s.values.push(r)):this.observers.set(t,{observed:1,values:[r]}),r}static observer(t){return this.observers.get(t)||null}static serialize(t){return JSON.stringify(t)}}e.SignalRegistry=o,o.observers=new Map,e.default=function(t){if(!n.default.current)throw i.default.for("Signals can only be called within a Temple component");if(!n.default.current.initiated)return o.observe(n.default.current,t);const e=o.observer(n.default.current);if(!e)throw i.default.for("State mismatch");return e.values[e.observed++%e.values.length]}},552:function(t,e,r){"use strict";var s=this&&this.__createBinding||(Object.create?function(t,e,r,s){void 0===s&&(s=r);var n=Object.getOwnPropertyDescriptor(e,r);n&&!("get"in n?!e.__esModule:n.writable||n.configurable)||(n={enumerable:!0,get:function(){return e[r]}}),Object.defineProperty(t,s,n)}:function(t,e,r,s){void 0===s&&(s=r),t[s]=e[r]}),n=this&&this.__setModuleDefault||(Object.create?function(t,e){Object.defineProperty(t,"default",{enumerable:!0,value:e})}:function(t,e){t.default=e}),i=this&&this.__importStar||function(t){if(t&&t.__esModule)return t;var e={};if(null!=t)for(var r in t)"default"!==r&&Object.prototype.hasOwnProperty.call(t,r)&&s(e,t,r);return n(e,t),e},o=this&&this.__importDefault||function(t){return t&&t.__esModule?t:{default:t}};Object.defineProperty(e,"__esModule",{value:!0}),e.TempleComponent=e.SignalRegistry=e.signal=e.props=void 0;const a=o(r(536));e.TempleComponent=a.default;const u=o(r(357));e.props=u.default;const c=i(r(975));e.signal=c.default,Object.defineProperty(e,"SignalRegistry",{enumerable:!0,get:function(){return c.SignalRegistry}})},806:function(t,e,r){"use strict";var s=this&&this.__createBinding||(Object.create?function(t,e,r,s){void 0===s&&(s=r);var n=Object.getOwnPropertyDescriptor(e,r);n&&!("get"in n?!e.__esModule:n.writable||n.configurable)||(n={enumerable:!0,get:function(){return e[r]}}),Object.defineProperty(t,s,n)}:function(t,e,r,s){void 0===s&&(s=r),t[s]=e[r]}),n=this&&this.__exportStar||function(t,e){for(var r in t)"default"===r||Object.prototype.hasOwnProperty.call(e,r)||s(e,t,r)};Object.defineProperty(e,"__esModule",{value:!0}),n(r(552),e)},237:(t,e,r)=>{t.exports={...r(806)}},531:(t,e,r)=>{"use strict";Object.defineProperty(e,"__esModule",{value:!0});const s=r(339),n=r(536),i=r(237);class o extends n.default{styles(){return""}template(){const{start:t}=(0,i.props)(),e=(0,i.signal)(t);console.log("count",e,e.value);const r=()=>{console.log("up",e.value),e.value++};return()=>[...e.value>0?[document.createTextNode("\n  "),s.default.create("div",{},[document.createTextNode("\n  Reactive Counter: "),s.default.create("strong",{},[document.createTextNode(String(e.value))]).element,document.createTextNode("\n  ")]).element,document.createTextNode("\n")]:[],document.createTextNode("\n\n"),s.default.create("button",{click:r},[document.createTextNode("Add")]).element]}}e.default=o,customElements.define("temple-counter",o)},646:(t,e,r)=>{"use strict";Object.defineProperty(e,"__esModule",{value:!0});const s=r(339),n=r(536),i=r(237);class o extends n.default{styles(){return":host { font-size: 21px; }"}template(){const{className:t,children:e}=(0,i.props)();return console.log((0,i.props)()),()=>[s.default.create("h1",{class:t},[document.createTextNode(String(e))]).element]}}e.default=o,customElements.define("temple-header",o)},943:(t,e,r)=>{"use strict";Object.defineProperty(e,"__esModule",{value:!0});const s=r(339),n=r(536);class i extends n.default{styles(){return""}template(){const t=["a","b","c"];return()=>[...t.length>0?[document.createTextNode("\n  "),s.default.create("ul",{},[document.createTextNode("\n  "),...Object.entries(t).map((([t,e])=>[document.createTextNode("\n    "),s.default.create("li",{},[document.createTextNode(String(e))]).element,document.createTextNode("\n  ")])).flat(),document.createTextNode("\n  ")]).element,document.createTextNode("\n")]:[]]}}e.default=i,customElements.define("temple-list",i)},721:(t,e,r)=>{"use strict";Object.defineProperty(e,"__esModule",{value:!0});const s=r(339),n=r(536),i=r(237);class o extends n.default{styles(){return""}template(){const{classname:t,children:e}=(0,i.props)();return console.log((0,i.props)()),()=>[s.default.create("p",{},[document.createTextNode("\n  "),document.createTextNode(String(e)),document.createTextNode("\n  "),s.default.create("temple-redirect",{class:t,href:"/"},[document.createTextNode("Link")]).element,document.createTextNode("\n")]).element]}}e.default=o,customElements.define("temple-paragraph",o)},791:(t,e,r)=>{"use strict";Object.defineProperty(e,"__esModule",{value:!0});const s=r(339),n=r(536),i=r(237);class o extends n.default{styles(){return":host { font-size: 30px; }"}template(){const{href:t,classname:e,children:r}=(0,i.props)(),n={title:"Click me"};return()=>[s.default.create("a",{href:t,...n},[document.createTextNode(String(r))]).element]}}e.default=o,customElements.define("temple-redirect",o)}},e={};function r(s){var n=e[s];if(void 0!==n)return n.exports;var i=e[s]={exports:{}};return t[s].call(i.exports,i,i.exports,r),i.exports}(()=>{"use strict";r(214),r(646),r(791),r(721),r(943),r(531)})()})();</script></head><body>
    <temple-header class="title">This is Header 1</temple-header>
    <temple-paragraph className="description">This is Paragraph</temple-paragraph>
    <temple-list />
    <temple-counter start="prop:value" />
  </body></html>